Catch-all fallback for libs detection

HadrienG2 opened the pull request:
The `_libs_default_handler` implicitly assumes that a library package called `(lib)?Something` will always produce a library called `libSomething`. This is not always the case, here are some counter examples:

- `gettext` (produces several `libgettextSomething`s and a `libintl`, but not `libgettext`)
- `intel-Something` (produces `libSomething`, without the intel prefix)
- X11 packages (though that one could be fixed with a case-insensitive search)

Currently, the result is that these libraries will not be correctly added to the RPATH, leading to problems like #10617 . One possibility is to override libs() for every such package, but that gets tedious quickly. As an alternative, this PR proposes to have the default libs handler try harder and enumerate every library in the package's Spack prefix instead.

I am quite convinced that this is a decent heuristic for library packages that are build using Spack, as the prefix will be specific to that package. But am not sure if this heuristic works or can be made to work for external packages where many libraries can end up sharing a common prefix. Extra review on this specific point would be most welcome.

See also #10617 for prior discussion.

davydden commented:
> One possibility is to override libs() for every such package, but that gets tedious quickly.

another question to ask ourselves: how many packages are non-standard? If only 10-15 our of 2000, then one might as well provide a custom `libs()` and call it a day.

But regardless, let's wait for others to comment.

HadrienG2 commented:
> another question to ask ourselves: how many packages are non-standard? If only 10-15 our of 2000, then one might as well provide a custom `libs()` and call it a day.

I can answer this question for the set of 98 packages that I have built recently, but maybe someone has a larger Spack package set to run this experiment on?

HadrienG2 commented:
So, here is the answer to @davydden on my Spack installation. Using the following Python script...

```python
#!/usr/bin/env python3
#
# Run this script in your Spack prefix to evaluate how badly the
# _libs_default_handler would mess with your library packages.

import os
import re

# List all folders which contain libraries, and libraries inside each of them
lib_list = []
lib_regex = re.compile(r"^lib[^.]*\.(l?a|so)")
for path, _dirs, files in os.walk('opt/spack'):
    lib_file_list = None
    for file in files:
        if lib_regex.match(file):
            if lib_file_list is None:
                lib_list.append((path, []))
                lib_file_list = lib_list[-1][1]
            lib_file_list.append(file)

# Check which library paths would be excluded by the default libs handler
altered_lib_paths = []
discarded_lib_paths = []
for lib_path, lib_files in lib_list:
    # Extract the package name
    package_version = lib_path.split('/')[4]
    package = '-'.join(package_version.split('-')[:-2])
    
    # Build a regex matching the default libs handler's name expectations
    package_libname = package
    if not package_libname.startswith('lib'):
        package_libname = 'lib' + package
    package_libname_re = re.compile("^"+package_libname.replace('-', '.'))

    # Check which libraries (or library paths!) don't match this regex
    altered_path = False
    discarded_path = True
    for lib_name in lib_files:
        if not package_libname_re.match(lib_name):
            if not altered_path:
                print('- In library path', lib_path, 'from package', package)
                altered_lib_paths.append(lib_path)
                altered_path = True
            print('  * Library', lib_name, 'would be excluded from libs()')
        else:
            discarded_path = False

    # Would this be recognized as a library?
    if discarded_path:
        print('  * This would exclude the entire library path from libs()!')
        discarded_lib_paths.append(lib_path)

# Print our conclusions
print('Of', len(lib_list), 'library paths found in the Spack prefix...')
print('-', len(altered_lib_paths), 'would have missing libraries in libs()')
print('-', len(discarded_lib_paths), 'would be discarded from libs() entirely')
```

I established that among my 98 Spack packages, there are 66 library paths. 40 of these would be altered by the current default libs() handler, and 34 would be discarded entirely.

Note that the heuristic used by this script is imperfect. It may have false negatives and false positives here and there. However, looking at the logs, most detected failures of the default libs() handler seem legit to me:

```commandline
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/automake-1.16.1-akaublt7ptqay3ty3xwvse7rnpbh57xv/share/automake-1.16/am from package automake
  * Library library.am would be excluded from libs()
  * Library libs.am would be excluded from libs()
  * Library libtool.am would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/binutils-2.31.1-4vrebnglbfxiuoevdfilbvb7uzadvqxg/lib from package binutils
  * Library libbfd.a would be excluded from libs()
  * Library libbfd.la would be excluded from libs()
  * Library libbfd.so would be excluded from libs()
  * Library libopcodes.a would be excluded from libs()
  * Library libopcodes.la would be excluded from libs()
  * Library libopcodes.so would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/bison-3.0.5-jdhs3kma4bfgoovzu7mmaqvvwjr7x5rp/lib from package bison
  * Library liby.a would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/bzip2-1.0.6-4c2n2owldmqp7kb25m7l7tyuq27abx76/lib from package bzip2
  * Library libbz2.a would be excluded from libs()
  * Library libbz2.so would be excluded from libs()
  * Library libbz2.so.1 would be excluded from libs()
  * Library libbz2.so.1.0 would be excluded from libs()
  * Library libbz2.so.1.0.6 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/flex-2.6.3-c3c4ufnpevtmf6wi3kucvihmwq3n3ucd/lib from package flex
  * Library libfl.a would be excluded from libs()
  * Library libfl.la would be excluded from libs()
  * Library libfl.so would be excluded from libs()
  * Library libfl.so.2 would be excluded from libs()
  * Library libfl.so.2.0.0 would be excluded from libs()
  * Library libl.a would be excluded from libs()
  * Library libl.so would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/gettext-0.19.8.1-vs5xjy6g6nwkk6dgnyv3wshqirejros2/lib from package gettext
  * Library libasprintf.a would be excluded from libs()
  * Library libasprintf.la would be excluded from libs()
  * Library libasprintf.so would be excluded from libs()
  * Library libasprintf.so.0 would be excluded from libs()
  * Library libasprintf.so.0.0.0 would be excluded from libs()
  * Library libintl.a would be excluded from libs()
  * Library libintl.la would be excluded from libs()
  * Library libintl.so would be excluded from libs()
  * Library libintl.so.8 would be excluded from libs()
  * Library libintl.so.8.1.5 would be excluded from libs()
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/glew-2.0.0-tchmig4vn4pc3ffe5mstohhylpnjl6gq/lib64 from package glew
  * Library libGLEW.a would be excluded from libs()
  * Library libGLEW.so would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/icu4c-60.1-jwsrqx7awhj43avj5dknrvoa7weh6uq4/lib from package icu4c
  * Library libicudata.so would be excluded from libs()
  * Library libicudata.so.60 would be excluded from libs()
  * Library libicudata.so.60.1 would be excluded from libs()
  * Library libicui18n.so would be excluded from libs()
  * Library libicui18n.so.60 would be excluded from libs()
  * Library libicui18n.so.60.1 would be excluded from libs()
  * Library libicuio.so would be excluded from libs()
  * Library libicuio.so.60 would be excluded from libs()
  * Library libicuio.so.60.1 would be excluded from libs()
  * Library libicutest.so would be excluded from libs()
  * Library libicutest.so.60 would be excluded from libs()
  * Library libicutest.so.60.1 would be excluded from libs()
  * Library libicutu.so would be excluded from libs()
  * Library libicutu.so.60 would be excluded from libs()
  * Library libicutu.so.60.1 would be excluded from libs()
  * Library libicuuc.so would be excluded from libs()
  * Library libicuuc.so.60 would be excluded from libs()
  * Library libicuuc.so.60.1 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/intel-tbb-2019.2-43mkxnfrcyrf5ohameeyzowchwgj2abr/lib from package intel-tbb
  * Library libtbb.so would be excluded from libs()
  * Library libtbb.so.2 would be excluded from libs()
  * Library libtbb_debug.so would be excluded from libs()
  * Library libtbb_debug.so.2 would be excluded from libs()
  * Library libtbbmalloc.so would be excluded from libs()
  * Library libtbbmalloc.so.2 would be excluded from libs()
  * Library libtbbmalloc_debug.so would be excluded from libs()
  * Library libtbbmalloc_debug.so.2 would be excluded from libs()
  * Library libtbbmalloc_proxy.so would be excluded from libs()
  * Library libtbbmalloc_proxy.so.2 would be excluded from libs()
  * Library libtbbmalloc_proxy_debug.so would be excluded from libs()
  * Library libtbbmalloc_proxy_debug.so.2 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libdrm-2.4.81-67eiywg72mzn6gxksugr6n5a3yudjzvc/lib from package libdrm
  * Library libkms.a would be excluded from libs()
  * Library libkms.la would be excluded from libs()
  * Library libkms.so would be excluded from libs()
  * Library libkms.so.1 would be excluded from libs()
  * Library libkms.so.1.0.0 would be excluded from libs()
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libice-1.0.9-gaz75ci343uhf5td3dlhroc2ujl6jy5v/lib from package libice
  * Library libICE.a would be excluded from libs()
  * Library libICE.la would be excluded from libs()
  * Library libICE.so would be excluded from libs()
  * Library libICE.so.6 would be excluded from libs()
  * Library libICE.so.6.3.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libiconv-1.15-fk4e54f7vilb7g63jn63kiv2w6kwu46x/lib from package libiconv
  * Library libcharset.a would be excluded from libs()
  * Library libcharset.la would be excluded from libs()
  * Library libcharset.so would be excluded from libs()
  * Library libcharset.so.1 would be excluded from libs()
  * Library libcharset.so.1.0.0 would be excluded from libs()
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libsm-1.2.2-o5if5pwxf24cz57z7hywzkdv4qxs6li3/lib from package libsm
  * Library libSM.a would be excluded from libs()
  * Library libSM.la would be excluded from libs()
  * Library libSM.so would be excluded from libs()
  * Library libSM.so.6 would be excluded from libs()
  * Library libSM.so.6.0.1 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libtool-2.4.6-6nyrrltokhmtbgefznc3bgu3fn6yuyjo/lib from package libtool
  * Library libltdl.a would be excluded from libs()
  * Library libltdl.la would be excluded from libs()
  * Library libltdl.so would be excluded from libs()
  * Library libltdl.so.7 would be excluded from libs()
  * Library libltdl.so.7.3.1 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libx11-1.6.7-xx3xvuxy2wqufwffz7566rhrcnipgwon/lib from package libx11
  * Library libX11-xcb.a would be excluded from libs()
  * Library libX11-xcb.la would be excluded from libs()
  * Library libX11-xcb.so would be excluded from libs()
  * Library libX11-xcb.so.1 would be excluded from libs()
  * Library libX11-xcb.so.1.0.0 would be excluded from libs()
  * Library libX11.a would be excluded from libs()
  * Library libX11.la would be excluded from libs()
  * Library libX11.so would be excluded from libs()
  * Library libX11.so.6 would be excluded from libs()
  * Library libX11.so.6.3.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxau-1.0.8-epl6g6e4obagkzs3vs6grhokxavgd7wm/lib from package libxau
  * Library libXau.a would be excluded from libs()
  * Library libXau.la would be excluded from libs()
  * Library libXau.so would be excluded from libs()
  * Library libXau.so.6 would be excluded from libs()
  * Library libXau.so.6.0.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxdamage-1.1.4-lpc44ywdsz3ohciwycwdcaiadhusn2gh/lib from package libxdamage
  * Library libXdamage.a would be excluded from libs()
  * Library libXdamage.la would be excluded from libs()
  * Library libXdamage.so would be excluded from libs()
  * Library libXdamage.so.1 would be excluded from libs()
  * Library libXdamage.so.1.1.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxdmcp-1.1.2-xypltxmzk7ggw7tqhqbgjvquhznsjx5o/lib from package libxdmcp
  * Library libXdmcp.a would be excluded from libs()
  * Library libXdmcp.la would be excluded from libs()
  * Library libXdmcp.so would be excluded from libs()
  * Library libXdmcp.so.6 would be excluded from libs()
  * Library libXdmcp.so.6.0.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxext-1.3.3-s5oivdt2e7wqkav7xln5so5xcx5kfwng/lib from package libxext
  * Library libXext.a would be excluded from libs()
  * Library libXext.la would be excluded from libs()
  * Library libXext.so would be excluded from libs()
  * Library libXext.so.6 would be excluded from libs()
  * Library libXext.so.6.4.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxfixes-5.0.2-57z722ldid7hbzcxmmelhbz7gjwaywyx/lib from package libxfixes
  * Library libXfixes.a would be excluded from libs()
  * Library libXfixes.la would be excluded from libs()
  * Library libXfixes.so would be excluded from libs()
  * Library libXfixes.so.3 would be excluded from libs()
  * Library libXfixes.so.3.1.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxft-2.3.2-5k46lczfgxxq74dlcu75xk7rhw43sgna/lib from package libxft
  * Library libXft.a would be excluded from libs()
  * Library libXft.la would be excluded from libs()
  * Library libXft.so would be excluded from libs()
  * Library libXft.so.2 would be excluded from libs()
  * Library libXft.so.2.3.2 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxi-1.7.6-wi6g2stajmtbbjpazcpkut33npnxdcce/lib from package libxi
  * Library libXi.a would be excluded from libs()
  * Library libXi.la would be excluded from libs()
  * Library libXi.so would be excluded from libs()
  * Library libXi.so.6 would be excluded from libs()
  * Library libXi.so.6.1.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxmu-1.1.2-etmfc3hrc3tljqxgaacj2ohbqkhj6qze/lib from package libxmu
  * Library libXmu.a would be excluded from libs()
  * Library libXmu.la would be excluded from libs()
  * Library libXmu.so would be excluded from libs()
  * Library libXmu.so.6 would be excluded from libs()
  * Library libXmu.so.6.2.0 would be excluded from libs()
  * Library libXmuu.a would be excluded from libs()
  * Library libXmuu.la would be excluded from libs()
  * Library libXmuu.so would be excluded from libs()
  * Library libXmuu.so.1 would be excluded from libs()
  * Library libXmuu.so.1.0.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxpm-3.5.12-ounjnw5avvbm3zjbv76ekixjvvjmw2gt/lib from package libxpm
  * Library libXpm.a would be excluded from libs()
  * Library libXpm.la would be excluded from libs()
  * Library libXpm.so would be excluded from libs()
  * Library libXpm.so.4 would be excluded from libs()
  * Library libXpm.so.4.11.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxrender-0.9.10-7egiuwglxvqwdavaxpqduh6dajmk3cmc/lib from package libxrender
  * Library libXrender.a would be excluded from libs()
  * Library libXrender.la would be excluded from libs()
  * Library libXrender.so would be excluded from libs()
  * Library libXrender.so.1 would be excluded from libs()
  * Library libXrender.so.1.3.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxt-1.1.5-lxhs7dphw2n4vkniu44hriyeissxsup7/lib from package libxt
  * Library libXt.a would be excluded from libs()
  * Library libXt.la would be excluded from libs()
  * Library libXt.so would be excluded from libs()
  * Library libXt.so.6 would be excluded from libs()
  * Library libXt.so.6.0.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxv-1.0.10-zxlg7kymi3iepob4lzpdcn2gcq72wcno/lib from package libxv
  * Library libXv.a would be excluded from libs()
  * Library libXv.la would be excluded from libs()
  * Library libXv.so would be excluded from libs()
  * Library libXv.so.1 would be excluded from libs()
  * Library libXv.so.1.0.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxvmc-1.0.9-2bqbh7grnmwadguneu5vcc2gvythzj4j/lib from package libxvmc
  * Library libXvMC.a would be excluded from libs()
  * Library libXvMC.la would be excluded from libs()
  * Library libXvMC.so would be excluded from libs()
  * Library libXvMC.so.1 would be excluded from libs()
  * Library libXvMC.so.1.0.0 would be excluded from libs()
  * Library libXvMCW.a would be excluded from libs()
  * Library libXvMCW.la would be excluded from libs()
  * Library libXvMCW.so would be excluded from libs()
  * Library libXvMCW.so.1 would be excluded from libs()
  * Library libXvMCW.so.1.0.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/libxxf86vm-1.1.4-gumjoejbfwnwpsd4wg6tvm7x4emqi4fs/lib from package libxxf86vm
  * Library libXxf86vm.a would be excluded from libs()
  * Library libXxf86vm.la would be excluded from libs()
  * Library libXxf86vm.so would be excluded from libs()
  * Library libXxf86vm.so.1 would be excluded from libs()
  * Library libXxf86vm.so.1.0.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/mesa-18.1.2-ze2efsp4kz65taeteebie5ger7xr4ozh/lib from package mesa
  * Library libGL.la would be excluded from libs()
  * Library libGL.so would be excluded from libs()
  * Library libGL.so.1 would be excluded from libs()
  * Library libGL.so.1.5.0 would be excluded from libs()
  * Library libGLESv1_CM.la would be excluded from libs()
  * Library libGLESv1_CM.so would be excluded from libs()
  * Library libGLESv1_CM.so.1 would be excluded from libs()
  * Library libGLESv1_CM.so.1.1.0 would be excluded from libs()
  * Library libGLESv2.la would be excluded from libs()
  * Library libGLESv2.so would be excluded from libs()
  * Library libGLESv2.so.2 would be excluded from libs()
  * Library libGLESv2.so.2.0.0 would be excluded from libs()
  * Library libOSMesa.la would be excluded from libs()
  * Library libOSMesa.so would be excluded from libs()
  * Library libOSMesa.so.8 would be excluded from libs()
  * Library libOSMesa.so.8.0.0 would be excluded from libs()
  * Library libglapi.la would be excluded from libs()
  * Library libglapi.so would be excluded from libs()
  * Library libglapi.so.0 would be excluded from libs()
  * Library libglapi.so.0.0.0 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/mesa-glu-9.0.0-lixw5as25prrn4lhkxtagma7u36j4edu/lib from package mesa-glu
  * Library libGLU.a would be excluded from libs()
  * Library libGLU.la would be excluded from libs()
  * Library libGLU.so would be excluded from libs()
  * Library libGLU.so.1 would be excluded from libs()
  * Library libGLU.so.1.3.1 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/ncurses-6.1-5gpxxdfqx53j7vod5vhmjyameqkqxuan/lib from package ncurses
  * Library libcurses.a would be excluded from libs()
  * Library libcurses.so would be excluded from libs()
  * Library libform.a would be excluded from libs()
  * Library libform.so would be excluded from libs()
  * Library libform.so.6 would be excluded from libs()
  * Library libform.so.6.1 would be excluded from libs()
  * Library libform_g.a would be excluded from libs()
  * Library libformw.a would be excluded from libs()
  * Library libformw.so would be excluded from libs()
  * Library libformw.so.6 would be excluded from libs()
  * Library libformw.so.6.1 would be excluded from libs()
  * Library libformw_g.a would be excluded from libs()
  * Library libmenu.a would be excluded from libs()
  * Library libmenu.so would be excluded from libs()
  * Library libmenu.so.6 would be excluded from libs()
  * Library libmenu.so.6.1 would be excluded from libs()
  * Library libmenu_g.a would be excluded from libs()
  * Library libmenuw.a would be excluded from libs()
  * Library libmenuw.so would be excluded from libs()
  * Library libmenuw.so.6 would be excluded from libs()
  * Library libmenuw.so.6.1 would be excluded from libs()
  * Library libmenuw_g.a would be excluded from libs()
  * Library libpanel.a would be excluded from libs()
  * Library libpanel.so would be excluded from libs()
  * Library libpanel.so.6 would be excluded from libs()
  * Library libpanel.so.6.1 would be excluded from libs()
  * Library libpanel_g.a would be excluded from libs()
  * Library libpanelw.a would be excluded from libs()
  * Library libpanelw.so would be excluded from libs()
  * Library libpanelw.so.6 would be excluded from libs()
  * Library libpanelw.so.6.1 would be excluded from libs()
  * Library libpanelw_g.a would be excluded from libs()
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/nettle-3.4-cqodfzqgc5hfjiom3v7rws5r7fg7mkxn/lib64 from package nettle
  * Library libhogweed.a would be excluded from libs()
  * Library libhogweed.so would be excluded from libs()
  * Library libhogweed.so.4 would be excluded from libs()
  * Library libhogweed.so.4.4 would be excluded from libs()
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/openssl-1.1.1-w6tuzkujip7rbzestp6rgs3gowhx5ua7/lib from package openssl
  * Library libcrypto.a would be excluded from libs()
  * Library libcrypto.so would be excluded from libs()
  * Library libcrypto.so.1.1 would be excluded from libs()
  * Library libssl.a would be excluded from libs()
  * Library libssl.so would be excluded from libs()
  * Library libssl.so.1.1 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/readline-7.0-6ke6lu2nuq4p6nh5kp4npt6dcuvvyc7a/lib from package readline
  * Library libhistory.a would be excluded from libs()
  * Library libhistory.so would be excluded from libs()
  * Library libhistory.so.7 would be excluded from libs()
  * Library libhistory.so.7.0 would be excluded from libs()
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/root-6.14.04-2yrghrs42pfn4t7wgotsanzoa5tkbamo/lib from package root
  * Library libASImage.so would be excluded from libs()
  * Library libASImage.so.6.14 would be excluded from libs()
  * Library libASImage.so.6.14.04 would be excluded from libs()
  * Library libASImageGui.so would be excluded from libs()
  * Library libASImageGui.so.6.14 would be excluded from libs()
  * Library libASImageGui.so.6.14.04 would be excluded from libs()
  * Library libCling.so would be excluded from libs()
  * Library libCling.so.6.14 would be excluded from libs()
  * Library libCling.so.6.14.04 would be excluded from libs()
  * Library libCore.so would be excluded from libs()
  * Library libCore.so.6.14 would be excluded from libs()
  * Library libCore.so.6.14.04 would be excluded from libs()
  * Library libEG.so would be excluded from libs()
  * Library libEG.so.6.14 would be excluded from libs()
  * Library libEG.so.6.14.04 would be excluded from libs()
  * Library libEve.so would be excluded from libs()
  * Library libEve.so.6.14 would be excluded from libs()
  * Library libEve.so.6.14.04 would be excluded from libs()
  * Library libFTGL.so would be excluded from libs()
  * Library libFTGL.so.6.14 would be excluded from libs()
  * Library libFTGL.so.6.14.04 would be excluded from libs()
  * Library libFitPanel.so would be excluded from libs()
  * Library libFitPanel.so.6.14 would be excluded from libs()
  * Library libFitPanel.so.6.14.04 would be excluded from libs()
  * Library libFoam.so would be excluded from libs()
  * Library libFoam.so.6.14 would be excluded from libs()
  * Library libFoam.so.6.14.04 would be excluded from libs()
  * Library libFumili.so would be excluded from libs()
  * Library libFumili.so.6.14 would be excluded from libs()
  * Library libFumili.so.6.14.04 would be excluded from libs()
  * Library libGX11.so would be excluded from libs()
  * Library libGX11.so.6.14 would be excluded from libs()
  * Library libGX11.so.6.14.04 would be excluded from libs()
  * Library libGX11TTF.so would be excluded from libs()
  * Library libGX11TTF.so.6.14 would be excluded from libs()
  * Library libGX11TTF.so.6.14.04 would be excluded from libs()
  * Library libGdml.so would be excluded from libs()
  * Library libGdml.so.6.14 would be excluded from libs()
  * Library libGdml.so.6.14.04 would be excluded from libs()
  * Library libGed.so would be excluded from libs()
  * Library libGed.so.6.14 would be excluded from libs()
  * Library libGed.so.6.14.04 would be excluded from libs()
  * Library libGenVector.so would be excluded from libs()
  * Library libGenVector.so.6.14 would be excluded from libs()
  * Library libGenVector.so.6.14.04 would be excluded from libs()
  * Library libGeom.so would be excluded from libs()
  * Library libGeom.so.6.14 would be excluded from libs()
  * Library libGeom.so.6.14.04 would be excluded from libs()
  * Library libGeomBuilder.so would be excluded from libs()
  * Library libGeomBuilder.so.6.14 would be excluded from libs()
  * Library libGeomBuilder.so.6.14.04 would be excluded from libs()
  * Library libGeomPainter.so would be excluded from libs()
  * Library libGeomPainter.so.6.14 would be excluded from libs()
  * Library libGeomPainter.so.6.14.04 would be excluded from libs()
  * Library libGpad.so would be excluded from libs()
  * Library libGpad.so.6.14 would be excluded from libs()
  * Library libGpad.so.6.14.04 would be excluded from libs()
  * Library libGraf.so would be excluded from libs()
  * Library libGraf.so.6.14 would be excluded from libs()
  * Library libGraf.so.6.14.04 would be excluded from libs()
  * Library libGraf3d.so would be excluded from libs()
  * Library libGraf3d.so.6.14 would be excluded from libs()
  * Library libGraf3d.so.6.14.04 would be excluded from libs()
  * Library libGui.so would be excluded from libs()
  * Library libGui.so.6.14 would be excluded from libs()
  * Library libGui.so.6.14.04 would be excluded from libs()
  * Library libGuiBld.so would be excluded from libs()
  * Library libGuiBld.so.6.14 would be excluded from libs()
  * Library libGuiBld.so.6.14.04 would be excluded from libs()
  * Library libGuiHtml.so would be excluded from libs()
  * Library libGuiHtml.so.6.14 would be excluded from libs()
  * Library libGuiHtml.so.6.14.04 would be excluded from libs()
  * Library libGviz3d.so would be excluded from libs()
  * Library libGviz3d.so.6.14 would be excluded from libs()
  * Library libGviz3d.so.6.14.04 would be excluded from libs()
  * Library libHist.so would be excluded from libs()
  * Library libHist.so.6.14 would be excluded from libs()
  * Library libHist.so.6.14.04 would be excluded from libs()
  * Library libHistPainter.so would be excluded from libs()
  * Library libHistPainter.so.6.14 would be excluded from libs()
  * Library libHistPainter.so.6.14.04 would be excluded from libs()
  * Library libHtml.so would be excluded from libs()
  * Library libHtml.so.6.14 would be excluded from libs()
  * Library libHtml.so.6.14.04 would be excluded from libs()
  * Library libImt.so would be excluded from libs()
  * Library libImt.so.6.14 would be excluded from libs()
  * Library libImt.so.6.14.04 would be excluded from libs()
  * Library libJupyROOT.so would be excluded from libs()
  * Library libJupyROOT.so.6.14 would be excluded from libs()
  * Library libJupyROOT.so.6.14.04 would be excluded from libs()
  * Library libMLP.so would be excluded from libs()
  * Library libMLP.so.6.14 would be excluded from libs()
  * Library libMLP.so.6.14.04 would be excluded from libs()
  * Library libMathCore.so would be excluded from libs()
  * Library libMathCore.so.6.14 would be excluded from libs()
  * Library libMathCore.so.6.14.04 would be excluded from libs()
  * Library libMathMore.so would be excluded from libs()
  * Library libMathMore.so.6.14 would be excluded from libs()
  * Library libMathMore.so.6.14.04 would be excluded from libs()
  * Library libMatrix.so would be excluded from libs()
  * Library libMatrix.so.6.14 would be excluded from libs()
  * Library libMatrix.so.6.14.04 would be excluded from libs()
  * Library libMinuit.so would be excluded from libs()
  * Library libMinuit.so.6.14 would be excluded from libs()
  * Library libMinuit.so.6.14.04 would be excluded from libs()
  * Library libMinuit2.so would be excluded from libs()
  * Library libMinuit2.so.6.14 would be excluded from libs()
  * Library libMinuit2.so.6.14.04 would be excluded from libs()
  * Library libMultiProc.so would be excluded from libs()
  * Library libMultiProc.so.6.14 would be excluded from libs()
  * Library libMultiProc.so.6.14.04 would be excluded from libs()
  * Library libNet.so would be excluded from libs()
  * Library libNet.so.6.14 would be excluded from libs()
  * Library libNet.so.6.14.04 would be excluded from libs()
  * Library libNew.so would be excluded from libs()
  * Library libNew.so.6.14 would be excluded from libs()
  * Library libNew.so.6.14.04 would be excluded from libs()
  * Library libPhysics.so would be excluded from libs()
  * Library libPhysics.so.6.14 would be excluded from libs()
  * Library libPhysics.so.6.14.04 would be excluded from libs()
  * Library libPostscript.so would be excluded from libs()
  * Library libPostscript.so.6.14 would be excluded from libs()
  * Library libPostscript.so.6.14.04 would be excluded from libs()
  * Library libProof.so would be excluded from libs()
  * Library libProof.so.6.14 would be excluded from libs()
  * Library libProof.so.6.14.04 would be excluded from libs()
  * Library libProofBench.so would be excluded from libs()
  * Library libProofBench.so.6.14 would be excluded from libs()
  * Library libProofBench.so.6.14.04 would be excluded from libs()
  * Library libProofDraw.so would be excluded from libs()
  * Library libProofDraw.so.6.14 would be excluded from libs()
  * Library libProofDraw.so.6.14.04 would be excluded from libs()
  * Library libProofPlayer.so would be excluded from libs()
  * Library libProofPlayer.so.6.14 would be excluded from libs()
  * Library libProofPlayer.so.6.14.04 would be excluded from libs()
  * Library libPyROOT.so would be excluded from libs()
  * Library libPyROOT.so.6.14 would be excluded from libs()
  * Library libPyROOT.so.6.14.04 would be excluded from libs()
  * Library libQuadp.so would be excluded from libs()
  * Library libQuadp.so.6.14 would be excluded from libs()
  * Library libQuadp.so.6.14.04 would be excluded from libs()
  * Library libRGL.so would be excluded from libs()
  * Library libRGL.so.6.14 would be excluded from libs()
  * Library libRGL.so.6.14.04 would be excluded from libs()
  * Library libRHTTP.so would be excluded from libs()
  * Library libRHTTP.so.6.14 would be excluded from libs()
  * Library libRHTTP.so.6.14.04 would be excluded from libs()
  * Library libRHTTPSniff.so would be excluded from libs()
  * Library libRHTTPSniff.so.6.14 would be excluded from libs()
  * Library libRHTTPSniff.so.6.14.04 would be excluded from libs()
  * Library libRIO.so would be excluded from libs()
  * Library libRIO.so.6.14 would be excluded from libs()
  * Library libRIO.so.6.14.04 would be excluded from libs()
  * Library libROOTCanvasPainter.so would be excluded from libs()
  * Library libROOTCanvasPainter.so.6.14 would be excluded from libs()
  * Library libROOTCanvasPainter.so.6.14.04 would be excluded from libs()
  * Library libROOTDataFrame.so would be excluded from libs()
  * Library libROOTDataFrame.so.6.14 would be excluded from libs()
  * Library libROOTDataFrame.so.6.14.04 would be excluded from libs()
  * Library libROOTFitPanelv7.so would be excluded from libs()
  * Library libROOTFitPanelv7.so.6.14 would be excluded from libs()
  * Library libROOTFitPanelv7.so.6.14.04 would be excluded from libs()
  * Library libROOTGpadv7.so would be excluded from libs()
  * Library libROOTGpadv7.so.6.14 would be excluded from libs()
  * Library libROOTGpadv7.so.6.14.04 would be excluded from libs()
  * Library libROOTGraphicsPrimitives.so would be excluded from libs()
  * Library libROOTGraphicsPrimitives.so.6.14 would be excluded from libs()
  * Library libROOTGraphicsPrimitives.so.6.14.04 would be excluded from libs()
  * Library libROOTHistDraw.so would be excluded from libs()
  * Library libROOTHistDraw.so.6.14 would be excluded from libs()
  * Library libROOTHistDraw.so.6.14.04 would be excluded from libs()
  * Library libROOTVecOps.so would be excluded from libs()
  * Library libROOTVecOps.so.6.14 would be excluded from libs()
  * Library libROOTVecOps.so.6.14.04 would be excluded from libs()
  * Library libROOTWebDisplay.so would be excluded from libs()
  * Library libROOTWebDisplay.so.6.14 would be excluded from libs()
  * Library libROOTWebDisplay.so.6.14.04 would be excluded from libs()
  * Library libRSQLite.so would be excluded from libs()
  * Library libRSQLite.so.6.14 would be excluded from libs()
  * Library libRSQLite.so.6.14.04 would be excluded from libs()
  * Library libRecorder.so would be excluded from libs()
  * Library libRecorder.so.6.14 would be excluded from libs()
  * Library libRecorder.so.6.14.04 would be excluded from libs()
  * Library libRint.so would be excluded from libs()
  * Library libRint.so.6.14 would be excluded from libs()
  * Library libRint.so.6.14.04 would be excluded from libs()
  * Library libRootAuth.so would be excluded from libs()
  * Library libRootAuth.so.6.14 would be excluded from libs()
  * Library libRootAuth.so.6.14.04 would be excluded from libs()
  * Library libSPlot.so would be excluded from libs()
  * Library libSPlot.so.6.14 would be excluded from libs()
  * Library libSPlot.so.6.14.04 would be excluded from libs()
  * Library libSQLIO.so would be excluded from libs()
  * Library libSQLIO.so.6.14 would be excluded from libs()
  * Library libSQLIO.so.6.14.04 would be excluded from libs()
  * Library libSessionViewer.so would be excluded from libs()
  * Library libSessionViewer.so.6.14 would be excluded from libs()
  * Library libSessionViewer.so.6.14.04 would be excluded from libs()
  * Library libSmatrix.so would be excluded from libs()
  * Library libSmatrix.so.6.14 would be excluded from libs()
  * Library libSmatrix.so.6.14.04 would be excluded from libs()
  * Library libSpectrum.so would be excluded from libs()
  * Library libSpectrum.so.6.14 would be excluded from libs()
  * Library libSpectrum.so.6.14.04 would be excluded from libs()
  * Library libSpectrumPainter.so would be excluded from libs()
  * Library libSpectrumPainter.so.6.14 would be excluded from libs()
  * Library libSpectrumPainter.so.6.14.04 would be excluded from libs()
  * Library libSrvAuth.so would be excluded from libs()
  * Library libSrvAuth.so.6.14 would be excluded from libs()
  * Library libSrvAuth.so.6.14.04 would be excluded from libs()
  * Library libThread.so would be excluded from libs()
  * Library libThread.so.6.14 would be excluded from libs()
  * Library libThread.so.6.14.04 would be excluded from libs()
  * Library libTree.so would be excluded from libs()
  * Library libTree.so.6.14 would be excluded from libs()
  * Library libTree.so.6.14.04 would be excluded from libs()
  * Library libTreePlayer.so would be excluded from libs()
  * Library libTreePlayer.so.6.14 would be excluded from libs()
  * Library libTreePlayer.so.6.14.04 would be excluded from libs()
  * Library libTreeViewer.so would be excluded from libs()
  * Library libTreeViewer.so.6.14 would be excluded from libs()
  * Library libTreeViewer.so.6.14.04 would be excluded from libs()
  * Library libVMC.so would be excluded from libs()
  * Library libVMC.so.6.14 would be excluded from libs()
  * Library libVMC.so.6.14.04 would be excluded from libs()
  * Library libX3d.so would be excluded from libs()
  * Library libX3d.so.6.14 would be excluded from libs()
  * Library libX3d.so.6.14.04 would be excluded from libs()
  * Library libXMLIO.so would be excluded from libs()
  * Library libXMLIO.so.6.14 would be excluded from libs()
  * Library libXMLIO.so.6.14.04 would be excluded from libs()
  * Library libcomplexDict.so would be excluded from libs()
  * Library libcomplexDict.so.6.14 would be excluded from libs()
  * Library libcomplexDict.so.6.14.04 would be excluded from libs()
  * Library libdequeDict.so would be excluded from libs()
  * Library libdequeDict.so.6.14 would be excluded from libs()
  * Library libdequeDict.so.6.14.04 would be excluded from libs()
  * Library libforward_listDict.so would be excluded from libs()
  * Library libforward_listDict.so.6.14 would be excluded from libs()
  * Library libforward_listDict.so.6.14.04 would be excluded from libs()
  * Library liblistDict.so would be excluded from libs()
  * Library liblistDict.so.6.14 would be excluded from libs()
  * Library liblistDict.so.6.14.04 would be excluded from libs()
  * Library libmap2Dict.so would be excluded from libs()
  * Library libmap2Dict.so.6.14 would be excluded from libs()
  * Library libmap2Dict.so.6.14.04 would be excluded from libs()
  * Library libmapDict.so would be excluded from libs()
  * Library libmapDict.so.6.14 would be excluded from libs()
  * Library libmapDict.so.6.14.04 would be excluded from libs()
  * Library libmultimap2Dict.so would be excluded from libs()
  * Library libmultimap2Dict.so.6.14 would be excluded from libs()
  * Library libmultimap2Dict.so.6.14.04 would be excluded from libs()
  * Library libmultimapDict.so would be excluded from libs()
  * Library libmultimapDict.so.6.14 would be excluded from libs()
  * Library libmultimapDict.so.6.14.04 would be excluded from libs()
  * Library libmultisetDict.so would be excluded from libs()
  * Library libmultisetDict.so.6.14 would be excluded from libs()
  * Library libmultisetDict.so.6.14.04 would be excluded from libs()
  * Library libsetDict.so would be excluded from libs()
  * Library libsetDict.so.6.14 would be excluded from libs()
  * Library libsetDict.so.6.14.04 would be excluded from libs()
  * Library libunordered_mapDict.so would be excluded from libs()
  * Library libunordered_mapDict.so.6.14 would be excluded from libs()
  * Library libunordered_mapDict.so.6.14.04 would be excluded from libs()
  * Library libunordered_multimapDict.so would be excluded from libs()
  * Library libunordered_multimapDict.so.6.14 would be excluded from libs()
  * Library libunordered_multimapDict.so.6.14.04 would be excluded from libs()
  * Library libunordered_multisetDict.so would be excluded from libs()
  * Library libunordered_multisetDict.so.6.14 would be excluded from libs()
  * Library libunordered_multisetDict.so.6.14.04 would be excluded from libs()
  * Library libunordered_setDict.so would be excluded from libs()
  * Library libunordered_setDict.so.6.14 would be excluded from libs()
  * Library libunordered_setDict.so.6.14.04 would be excluded from libs()
  * Library libvalarrayDict.so would be excluded from libs()
  * Library libvalarrayDict.so.6.14 would be excluded from libs()
  * Library libvalarrayDict.so.6.14.04 would be excluded from libs()
  * Library libvectorDict.so would be excluded from libs()
  * Library libvectorDict.so.6.14 would be excluded from libs()
  * Library libvectorDict.so.6.14.04 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/xz-5.2.4-zivqc2hzzvkzttpptd7rj2vmzrpahye6/lib from package xz
  * Library liblzma.a would be excluded from libs()
  * Library liblzma.la would be excluded from libs()
  * Library liblzma.so would be excluded from libs()
  * Library liblzma.so.5 would be excluded from libs()
  * Library liblzma.so.5.2.4 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/zlib-1.2.11-lbjnq6wr6twtldkwgwouotgzsiqubk3j/lib from package zlib
  * Library libz.a would be excluded from libs()
  * Library libz.so would be excluded from libs()
  * Library libz.so.1 would be excluded from libs()
  * Library libz.so.1.2.11 would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/acts-core-develop-lf3xc5bxotcjp2vniajy33bfs5rjdvei/lib64 from package acts-core
  * Library libActsCore.so would be excluded from libs()
  * Library libActsCustomLogger.so would be excluded from libs()
  * Library libActsDD4hepPlugin.so would be excluded from libs()
  * Library libActsDigitizationPlugin.so would be excluded from libs()
  * Library libActsJsonPlugin.so would be excluded from libs()
  * Library libActsMaterialMappingPlugin.so would be excluded from libs()
  * Library libActsTGeoPlugin.so would be excluded from libs()
  * This would exclude the entire library path from libs()!
- In library path opt/spack/linux-opensuse_tumbleweed20190205-x86_64/gcc-8.2.1/dd4hep-1.10.0-nd2wmq5e63gqfxkpn3o4nnp4mk4jsini/lib from package dd4hep
  * Library libDD4hepGaudiPluginMgr.so would be excluded from libs()
  * Library libDD4hepGaudiPluginMgr.so.1.10 would be excluded from libs()
  * Library libDDAlign.so would be excluded from libs()
  * Library libDDAlign.so.1.10 would be excluded from libs()
  * Library libDDAlignPlugins.so would be excluded from libs()
  * Library libDDAlignPlugins.so.1.10 would be excluded from libs()
  * Library libDDCond.so would be excluded from libs()
  * Library libDDCond.so.1.10 would be excluded from libs()
  * Library libDDCondPlugins.so would be excluded from libs()
  * Library libDDCondPlugins.so.1.10 would be excluded from libs()
  * Library libDDCore.so would be excluded from libs()
  * Library libDDCore.so.1.10 would be excluded from libs()
  * Library libDDCorePlugins.so would be excluded from libs()
  * Library libDDCorePlugins.so.1.10 would be excluded from libs()
  * Library libDDDetectors.so would be excluded from libs()
  * Library libDDDetectors.so.1.10 would be excluded from libs()
  * Library libDDEvePlugins.so would be excluded from libs()
  * Library libDDEvePlugins.so.1.10 would be excluded from libs()
  * Library libDDGDMLPlugins.so would be excluded from libs()
  * Library libDDGDMLPlugins.so.1.10 would be excluded from libs()
  * Library libDDParsers.so would be excluded from libs()
  * Library libDDParsers.so.1.10 would be excluded from libs()
  * Library libDDPythonPlugins.so would be excluded from libs()
  * Library libDDPythonPlugins.so.1.10 would be excluded from libs()
  * Library libDDRec.so would be excluded from libs()
  * Library libDDRec.so.1.10 would be excluded from libs()
  * Library libDDRecPlugins.so would be excluded from libs()
  * Library libDDRecPlugins.so.1.10 would be excluded from libs()
  * This would exclude the entire library path from libs()!
Of 66 library paths found in the Spack prefix...
- 40 would have missing libraries in libs()
- 34 would be discarded from libs() entirely
```

adamjstewart commented:
> But am not sure if this heuristic works or can be made to work for external packages where many libraries can end up sharing a common prefix.

Overall I think your suggestion makes perfect sense... except in the case of external packages. This would definitely be a disaster for external packages, and would lead to a lot of difficult to diagnose build failures that aren't reproducible for others.

We could try to hack something where we only add `lib*` in the case of Spack-installed packages, but that adds to the complexity of the package.

I've been away from Spack for a while. Did we ever merge those PRs that suggested replacing `-Lprefix/lib` and `-Iprefix/include` with `libs.directories[0]` and `headers.directories[0]` in the compiler wrappers? If not, then this isn't such a big issue. It's really only when a package directly references `libs` or `headers` that this would be a problem.

davydden commented:
@adamjstewart 

> Did we ever merge those PRs that suggested replacing `-Lprefix/lib` and `-Iprefix/include` with `libs.directories[0]` and `headers.directories[0]` in the compiler wrappers?

we did #8136 and frankly I am happy that we did so. I am sure we can figure out issues like this one.

HadrienG2 commented:
> > But am not sure if this heuristic works or can be made to work for external packages where many libraries can end up sharing a common prefix.
> 
> Overall I think your suggestion makes perfect sense... except in the case of external packages. This would definitely be a disaster for external packages, and would lead to a lot of difficult to diagnose build failures that aren't reproducible for others.
> 
> We could try to hack something where we only add `lib*` in the case of Spack-installed packages, but that adds to the complexity of the package.

How would one detect if a package is internal or external in the default libs handler? Is that exposed by the spec object?

adamjstewart commented:
^ @alalazo would know

scheibelp commented:
> How would one detect if a package is internal or external in the default libs handler? Is that exposed by the spec object?

`Spec.external` stores whether the spec is external (so `self.spec.external` in the `Package`)

HadrienG2 commented:
@scheibelp Thanks for the explanation! Latest commit implements the policy suggested by @adamjstewart of only enabliing the catch-all fallback for internal packages.

alalazo commented:
I think we overlooked one detail. The previous heuristic looked for a single library (modulo the comment over `spec_name`), while the fallback here can potentially match multiple libraries. If we match more than one library the order in which they appear in the list might matter and can't be arbitrarily decided by the searching algorithm. What do people think of adding a check on whether we have multiple matches and raise an exception in case?

@HadrienG2 Can you write "realistic" unit tests for this, based on a couple of packages that are considered relevant? You can mock their installation directory using `tmpdir` or similar fixtures. Let me know if you need help for that.

HadrienG2 commented:
@alalazo I think I need to understand better why the order of libs() should matter to Spack.

The only circumstance I can think about is when passing libs arguments to a linker, because these things never underwent a proper redesign since the single-pass punch card era. However, this is not the job of spack, but that of the build system of dependent packages. Hand-written makefiles, CMake modules, pkgconfig, etc. which depend on libraries will take care of listing library dependencies in the right order, and all they need from spack is a list of library search paths to look into.

As far as I know, Spack does not try to replace these components, and therefore does not need to know the libraries' dependency order. Even if it did, it could not generate a proper dependency list anyway, because it does not know which libs a dependent package links to. All Spack could do is to provide a catch-all libs list, which would be hopelessly pessimistic for large modular frameworks like Qt or ROOT.

Can you explain me in which circumstance the order of spack's libs() list should matter ?

----

As for unit tests, considering that the [Spack developer documentation on them is empty](https://spack.readthedocs.io/en/latest/developer_guide.html#id3) and that I've never worked on a large Python project before, I could indeed use a quick tour of what is needed to add a new test to Spack, which useful facilities I should know about, and some pointers to similar existing tests that I could use as a template.

alalazo commented:
> The only circumstance I can think about is when passing libs arguments to a linker, because these things never underwent a proper redesign since the single-pass punch card era.

That's exactly the use case I had in mind. Spack doesn't try to be a substitute of the build-system underneath, but sometimes needs to know how to link to a library in a certain configuration. 

This is particularly true for hand-written Makefiles. In many cases the upstream authors suppose that a user should edit the Makefile to supply proper linking lines. You can check `Hdf5.libs` as an example of a provider and `quantum-espresso` or `netcdf` as an example of consumers of this kind of information.

> Even if it did, it could not generate a proper dependency list anyway, because it does not know which libs a dependent package links to.

No, but packages in Spack can reply to queries. The dependent makes a query, the package replies with the correct list of libraries. This also works through virtual dependencies, so you can say:
```python
mpi_fortran_libs = spec['mpi:cxx'].libs
```
in a dependent and you'll get the list of libraries that are needed to link to the (deprecated) C++ interface of the underlying MPI.

alalazo commented:
> I could indeed use a quick tour of what is needed to add a new test to Spack, which useful facilities I should know about, and some pointers to similar existing tests that I could use as a template.

Spack uses [pytest](https://docs.pytest.org/en/latest/contents.html) underneath. You can run the current unit tests on your platform by giving the following command:
```console
$ spack test
```
In `pytest` you use [fixtures](https://docs.pytest.org/en/latest/fixture.html) to set-up a proper state for a test. I think you might need to use built-in fixtures [that operate on files and directories](https://docs.pytest.org/en/latest/tmpdir.html) like `tmpdir` or `tmpdir_factories` to construct a mock installation tree for this PR. An example where I construct a fixture similar to what I think might be needed here is:

https://github.com/spack/spack/blob/b721338d5a15bb427c2b3dedac7f3b2412466c74/lib/spack/spack/test/conftest.py#L719-L741

while an example of a test in the same PR that uses the fixture is:

https://github.com/spack/spack/blob/b721338d5a15bb427c2b3dedac7f3b2412466c74/lib/spack/spack/test/llnl/util/filesystem.py#L220-L239

HadrienG2 commented:
@alalazo Thanks for the explanations. I see the use case for ordered libs() better now, but still think it is a bit too specialized to warrant penalizing library package writers over the lack thereof (and having them reverse-engineer and write a dependency-ordered list in their package.py).

A quick grep for "def libs(" in Spack's built-in repo fuels this suspicion, with only 47 packages over ~3000 overriding the default libs handler (as flawed as it may currently be).

Therefore, to answer your original question...

> What do people think of adding a check on whether we have multiple matches and raise an exception in case?

I think this should not be considered a fatal error, but at best an optional and off-by-default warning. This way, our message for package writers becomes the following:

> Spack has a default libs() implementation which will produce an unordered list of all libraries exported by the package. If you export at most one library, if the libraries you export are independent, or if you expect dependent packages to either hardcode their link flags or use a reasonably modern build system such as CMake, SCons or Meson, then you may safely rely on this default implementation.
>
> If, on the other hand, you belong to the minority of packages that care about library order because the libs() list is used in dependent packages in order to produce a linker command line, you will need to write your own libs() handler, which exposes the library list in dependency order.

What do you think?

(Meanwhile, I will start thinking about some good unit tests to add to this PR.)

alalazo commented:
> What do you think?

I would write something much shorter - as it might show up for multiple packages in a DAG - but I like the idea of a warning. This will still give us pointers to packages that potentially need a custom `libs` property, but won't make the build fail in cases where having a correct `libs` is not essential.

HadrienG2 commented:
Ah, sorry, I did not intend the earlier prose as something that is actually displayed by spack, but only as a user-side interpretation of the proposed Spack behaviour.

Here is something which would be more suitable as a CLI warning (which would only be displayed if a package provides multiple libraries):

"By default, package libraries are listed in an arbitrary order. If you intend to use the libs() query as input to a linker, you need to provide a dependency-ordered override."

And then, if needed, we can link to a documentation page containing a longer-form text explaining the issue in more detail:

- Why doesn't Spack provide a dependency-ordered library list? (I think that it could do so given enough ldd trickery, but that the implementation effort/complexity/overhead wouldn't be worthwhile for this relatively narrow use case)
- What are the cases where a dependency-ordered library list is necessary? (You only need this if 1/a package exports multiple inter-dependent libraries, 2/the linker command line for a dependent package is generated by Spack).

davydden commented:
@scheibelp @alalazo anything holds this PR? 

@scheibelp would like to merge this prior to https://github.com/spack/spack/pull/10628 , whereas  https://github.com/spack/spack/pull/9032 is blocked by #10628 .

HadrienG2 commented:
Two things holding it right now: could use some unit tests and a warning when multiple libs are found. Both can probably be postponed to a later PR if you're in a hurry.

davydden commented:
> Both can probably be postponed to a later PR if you're in a hurry.

I am not in a hurry, it’s just that the other PR is hanging since September 2018, whereas I prefer to have things done and out of the way.

HadrienG2 commented:
I think this version addresses all the comments that were made so far. Please feel free to re-check it.

@alalazo Instead of a realistic test of the default libs handler, which looked like it would require a rather complex test harness, I tried to follow the example of #10623 by extracting the catch-all library search logic into an `llnl.util.filesystem` routine and testing that routine in isolation. Let me know if you are happy with this.

HadrienG2 commented:
Seems the python 2.6 CI build fell victim of Travis' timeout though. You may want to restart it.

davydden commented:
> I think this version addresses all the comments that were made so far. Please feel free to re-check it.

@scheibelp @alalazo ping.

HadrienG2 commented:
Ping?

davydden commented:
> Ping?

@scheibelp ping?

davydden commented:
@HadrienG2 could you please resolve conflicts?

HadrienG2 commented:
@davydden Done!

davydden commented:
@scheibelp ping?

HadrienG2 commented:
This has been waiting long enough and I am tired of perpetually rebasing it, so I'll be closing it.

I'll keep the branch around for a while on my repo in case someone feels like taking it over.
